clingo: modify recipe for bootstrapping

alalazo opened the pull request:
fixes #22463 

Modifications:
- [x] infer Python spec for bootstrapping using the `sys` module instead of `external find` (faster)
- [x] give hints to find Python headers
- [x] avoid building the clingo app and libpyclingo when bootstrapping
- [x] hint compiler to be used and operating system

alalazo committed:
clingo: modify recipe for bootstrapping

Modifications:
- clingo builds with shared Python only
  if ^python+shared
- avoid building the clingo app
  for bootstrapping
- don't link to libpython when bootstrapping

alalazo committed:
Remove option that breaks on linux

alalazo committed:
Give more hints for the current Python

alalazo committed:
Disable CLINGO_BUILD_PY_SHARED for bootstrapping

alalazo committed:
bootstrapping: try to detect the current python from std library

This is much faster than calling external executables

alalazo committed:
Fix compatibility with Python 2.6

becker33 commented on lib/spack/spack/bootstrap.py:
```diff
@@ -20,17 +25,33 @@
-    externals = python_cls.determine_spec_details(
-        python_prefix, [os.path.basename(sys.executable)])
-    external_python = externals[0]
+    external_python = spec_for_current_python()
```
We're  losing a lot of information here compared to `determine_spec_details`. How confident are you that we won't actually need that information?

alalazo replied:
The information here permits to distinguish among the interpreters provided in the `manylinux1` image:
```console
[root@90e0a523cd26 /]# ls /opt/python/
cp27-cp27m  cp27-cp27mu  cp35-cp35m  cp36-cp36m  cp37-cp37m  cp38-cp38  cp39-cp39
```
which, to my understanding of [PEP-513](https://www.python.org/dev/peps/pep-0513/#the-manylinux1-policy) are the only one we need to care about for Python on `x86_64` and versions 2.7, 3.5-39. In principle I think Python 2.6 should be treated the same as Python 2.7 (and that is not accounted for in this PR) but I think that for that old Python version we may just stick to whatever build the CentOS 6 image pulls in via yum.

alalazo committed:
Give hints on which compiler and OS to use when bootstrapping

This change hints which compiler to use for bootstrapping clingo
(either GCC or Apple Clang on MacOS). On Cray platforms it also
hints to build for the frontend system, where software is meant
to be installed.

alalazo commented:
@bvanessen @jjellio The last commit in this PR 22dc58f should make compiler selection for bootstrapping more robust. It requires to have a valid GCC available on linux and cray.

cosmicexplorer commented on lib/spack/spack/bootstrap.py:
```diff
@@ -20,17 +27,33 @@
+      https://www.python.org/dev/peps/pep-0513/
+      https://stackoverflow.com/a/35801395/771663
+    """
+    version_str = '.'.join(str(x) for x in sys.version_info[:2])
```
There's a line in this file later in `make_module_available()`: `spec.constrain('^python@%d.%d' % sys.version_info[:2])`. Where is spack eventually using the spec we generate in this function?

alalazo replied:
> Where is spack eventually using the spec we generate in this function?

In the `spack_python_interpreter` context manager at line 56. That context manager is combined with others in the `ensure_bootstrap_configuration` context manager (see `asp.py` for where that one is used). That said it might be a good idea to use the spec returned here also in:
```
spec.constrain('^python@%d.%d' % sys.version_info[:2])
```
I'll check on that, thanks!

alalazo replied:
done in b0a61fb67f21d02bd31c01f3bf22216ee58e7df5

jjellio commented:
I can add that this worked for me on a Cray platform (rhel8 + CCE 11.0.3)
I've given a few notes in the referenced issue about potential pitfalls around static GCC libraries. (but that's something I can work around)

alalazo commented:
@becker33 Adding some timing data here for reference:
```console
$ spack python
Spack version 0.16.1
Python 3.8.5, Linux x86_64
>>> import sys
>>> import timeit
>>> import os.path
>>> import spack.spec
>>> python_cls = type(spack.spec.Spec('python').package)
>>> python_prefix = os.path.dirname(os.path.dirname(sys.executable))
>>> timeit.repeat(lambda: python_cls.determine_spec_details(python_prefix, [os.path.basename(sys.executable)]), number=1)
[0.4135378760001913, 0.4188740810004674, 0.40364057999977376, 0.3976841709991277, 0.41728885599877685]
>>> import spack.bootstrap
>>> timeit.repeat(spack.bootstrap.spec_for_current_python, number=1)
[7.970899969222955e-05, 2.762300027825404e-05, 2.0912999389111064e-05, 9.948000297299586e-06, 8.969998816610314e-06]
```

alalazo committed:
Use spec_for_current_python to constrain module requirement
